ngraph/core/reference/src/runtime/reference/einsum.cpp

```diff
@@ -0,0 +1,1134 @@
+                                            input_shape,
+                                            order,
+                                            output_shape,
+                                            element_type.size());
+
+        return output;
+    }
+
+    template <typename T>
+    void reduce_input(HostTensorVector& inputs,
+                      std::vector<std::string>& input_subscripts,
+                      const std::string& output_subscript,
+                      size_t input_ind)
+    {
+        // perform sanity check for arguments
+        auto num_inputs = inputs.size();
+        NGRAPH_CHECK(num_inputs == input_subscripts.size(), "Each input must have own subscript.");
+        NGRAPH_CHECK(input_ind < num_inputs, "Input index is out of range.");
+
+        const auto& input_ptr = inputs[input_ind];
+        const auto& input_subscript = input_subscripts[input_ind];
+        const auto input_shape = input_ptr->get_shape();
+
+        // compute output shape and axes to reduce
+        ngraph::Shape output_shape;
+        ngraph::AxisSet reduced_axes;
+        auto labels = ngraph::opset7::Einsum::extract_labels(input_subscripts[input_ind]);
+        auto label_dim_map =
+            compute_label_dim_map(input_ptr->get_partial_shape().rank(), input_subscript);
+        std::string new_input_subscript = "";
+        for (const auto& label : labels)
+        {
+            // check if the current label is met in the other input subscripts
+            // or the output subscript
+            bool is_dim_reduced =
+                is_dimension_reduced(input_subscripts, output_subscript, label, {input_ind});
+
+            NGRAPH_CHECK(label_dim_map.find(label) != label_dim_map.end());
+            auto label_dims = label_dim_map[label];
+
+            // if label is not met, dimension corresponding to the label is to reduce
+            if (is_dim_reduced)
+            {
+                reduced_axes.insert(label_dims.begin(), label_dims.end());
+            }
+            else
+            {
+                for (auto label_dim : label_dims)
+                {
+                    output_shape.push_back(input_shape[label_dim]);
+                }
+                new_input_subscript += label;
+            }
+        }
+
+        if (reduced_axes.size() == 0)
+        {
+            // there is no axis to reduce
+            return;
+        }
+
+        HostTensorPtr output_ptr = std::shared_ptr<HostTensor>(
+            new HostTensor(input_ptr->get_element_type(), output_shape));
+
+        ngraph::runtime::reference::sum<T>(input_ptr->get_data_ptr<T>(),
+                                           output_ptr->get_data_ptr<T>(),
+                                           input_shape,
+                                           reduced_axes,
+                                           false);
+
+        // update a vector of inputs and input subscripts
+        inputs[input_ind] = output_ptr;
```

Comment:
@ilyachur, please pay attention to memory management in this implementation. Now I create separate `HostTensor` for each intermediate result. `HostTensor` provides convenient way to incapsulate memory pointer and shape that I need for this implementation, however, it does not provide a convenient way to reuse it for another tensor with a new shape.

Reply:
We decided to re-use some tensors allocated for intermediate results for the next steps of the computation where it is possible.